packages/relay/src/Environment.js

```diff
@@ -19,26 +60,52 @@ export default function createEnvironment(accessToken: string = '') {
     networkHeaders.authorization = accessToken;
   }
 
-  const fetchQuery = async (operation, variables): Promise<FetcherResponse> => {
-    // TODO: fetch persisted queries instead (based on operation.id)
-    const response = await (await fetch('https://graphql.kiwi.com/', {
-      method: 'POST',
-      headers: networkHeaders,
-      body: JSON.stringify({
-        query: operation.text,
-        variables,
-      }),
-    })).json();
-
-    if (response.errors) {
-      response.errors.forEach(error => console.warn(error));
-    }
-
-    return response;
+  const fetchQuery = (
+    operation,
+    variables,
+    cacheConfig,
+  ): Promise<FetcherResponse> => {
+    return Observable.create(observer => {
+      if (cacheConfig.force === true) {
+        return fetchFromTheNetwork(
+          networkHeaders,
+          operation,
+          variables,
+          observer,
+        );
+      }
+
+      // try to read content from the cache and use it if possible
+      AsyncStorage.getItem(ASYNC_STORE_KEY)
+        .then(content => {
+          if (content !== null) {
+            store.publish(new RecordSource(JSON.parse(content)));
+            const operationSelector = createOperationSelector(
+              operation,
+              variables,
+            );
+            if (store.check(operationSelector.root)) {
+              // we have all data in the store to fulfill this query so let's
+              // load it from the memory first and call the API after that
+              // this will make the UI feel really fast
+              observer.next(store.lookup(operationSelector.root));
```

Comment:
So, here you don't use `observer.complete()` because you will still use the observer in the `finally` clause. Correct?

Reply:
Yes. The observer wouldn't accept `next` after calling `complete`. I will call `complete` only when I am sure I will not call `next` (after network response basically).